### ISSUE TITLE
Enforce Liskov substitution principle for subclass `__init__` method

### ISSUE THREAD

**Issue by Jasha10** [microsoft/pyright#4097]
Here's an example of class inheritance violating the Liskov principle:
```python
class Base:
    def __init__(self, param1: int, param2: bool) -> None:
        ...


class SubClass(Base):
    def __init__(self, param1: int) -> None:
        ...
```
Running pyright on this example does not yield any errors. 
I'd expect an error saying something like `Type "(param1: int) -> None" cannot be assigned to type "(param1: int, param2: bool) -> None"`.

```text
$ pyright --version
pyright 1.1.277
```

**Comment by erictraut** [comment]
Pyright intentionally excludes the `__init__` method is from LSP enforcement checks since it's very common in Python to do so. Mypy and other Python type checkers also exclude `__init__` in this case.

**Comment by Jasha10** [comment]
Got it. Thanks.

**Comment by Jasha10** [comment]
It would seem that other methods (not just `__init__`) are excluded from LSP enforcement. The following does not give any errors:
```python
class Base:
    def foo(self, param1: int, param2: bool) -> None:
        ...

class SubClass(Base):
    def foo(self, param1: int) -> None:
        ...
```

**Comment by rchiodo** [comment]
Really weird that this just came in as a feature request. See this issue here: https://github.com/microsoft/pyright/issues/4098

But if you set this:

```json
    "python.analysis.diagnosticSeverityOverrides": {
        "reportIncompatibleMethodOverride": "error"
    },
```

SubClass should show an error:

![image](https://user-images.githubusercontent.com/19672699/198140628-7dc0b500-26fe-4dd8-a5a1-dad8ded2cae6.png)

**Comment by Jasha10** [comment]
Thanks @rchiodo! That does the trick.
